Refactor Build Controller to use Informers

csrwng opened the pull request:
Collapses the build and build pod controller into a single controller using informers

csrwng commented:
[test]

csrwng commented:
@bparees I'm still working on unit tests, but wanted to get this up sooner than later; ptal

bparees commented on pkg/build/controller/build/build_controller.go:
```diff
@@ -0,0 +1,821 @@
+	return bc.createBuildPod(build)
+}
+
+func (bc *BuildController) createPodSpec(originalBuild *buildapi.Build, ref string) (*kapi.Pod, error) {
```
godoc

bparees commented on pkg/build/controller/build/build_controller.go:
```diff
@@ -0,0 +1,821 @@
+	update.setPodNameAnnotation(buildPod.Name)
+	update.setPhase(buildapi.BuildPhasePending)
+	update.setReason("")
+	update.setMessage("")
```
i don't think you can clear the reason/message here.  you're at risk of racing against the build pod.  if the build pod runs first and updates the build reason/message because it failed, you're going to wipe it out.

csrwng replied:
How would we handle something like:
https://bugzilla.redhat.com/show_bug.cgi?id=1454468

bparees replied:
i'm not sure how that bug exists since the buildpod controller clears the reason/message when it sees the pod enter the running state.

but the main point is the buildpod controller used to always get the build from the api server, so it knew (within a very small window at least) if the build had already been marked failed (in which case it did nothing) or not (in which case it updated the reason/message normally).

I think your build patch logic needs to somehow take into account "is the build i'm about to patch already marked failed?  if so, do not update the reason+message"

csrwng replied:
If we're changing the phase to pending then it's not failed (according to the cache). If you mean that we should ensure that the build we're patching is the version that we know of in the cache, then an option is to specify the ResourceVersion in the patch. The patch would fail if the ResourceVersion has changed (per @deads2k). Would kind of defeat the purpose of switching to patch though.

bparees replied:
that sounds like one solution, yes.  what i suggested i think is a more fine grained approach (you don't fail if the version doesn't match in general, you only fail if the cache is stale in one particular way).

you are also free to skip any of this checking if you're not updating any of:  the build state, the build reason, or the build message, since those are the fields where you might clobber a value that was set by the build pod directly.

bparees commented on pkg/build/controller/build/build_controller.go:
```diff
@@ -0,0 +1,821 @@
+	case kapi.PodPending:
+		if build.Status.Phase != buildapi.BuildPhasePending {
+			update.setReason("")
+			update.setMessage("")
```
similar issue here.  if the build pod has already failed and updated the build by the time you see this event, you're going to overwrite the failed message/reason.

bparees commented on pkg/build/controller/build/build_controller.go:
```diff
@@ -0,0 +1,821 @@
+	// A periodic relist will send update events for all known pods.
+	curPod := cur.(*kapi.Pod)
+	oldPod := old.(*kapi.Pod)
+	if curPod.ResourceVersion == oldPod.ResourceVersion {
```
comment for why it's ok to ignore pod relists here so people don't copy this pattern.

bparees commented on pkg/build/controller/build/creationstrategy.go:
```diff
@@ -0,0 +1,46 @@
+package build
```
filename:  podcreationstrategy.go

otherwise i'd expect this to create builds, given the package it's in.

bparees commented on pkg/build/controller/build/creationstrategy.go:
```diff
@@ -0,0 +1,46 @@
+	case build.Spec.Strategy.CustomStrategy != nil:
+		pod, err = f.customBuildStrategy.CreateBuildPod(build)
+	case build.Spec.Strategy.JenkinsPipelineStrategy != nil:
+		return nil, nil
```
it should be an error if we get here.

bparees commented on pkg/cmd/server/bootstrappolicy/controller_policy.go:
```diff
@@ -49,11 +49,12 @@ func init() {
-			rbac.NewRule("get").Groups(imageGroup, legacyImageGroup).Resources("imagestreams").RuleOrDie(),
-			rbac.NewRule("get", "list", "create", "delete").Groups(kapiGroup).Resources("pods").RuleOrDie(),
+			rbac.NewRule("get", "list").Groups(imageGroup, legacyImageGroup).Resources("imagestreams").RuleOrDie(),
+			rbac.NewRule("get", "list").Groups(imageGroup, legacyImageGroup).Resources("secrets").RuleOrDie(),
```
group looks wrong for secrets?

bparees commented on pkg/cmd/server/origin/controller/build.go:
```diff
@@ -7,8 +7,7 @@ import (
 	buildoverrides "github.com/openshift/origin/pkg/build/admission/overrides"
-	buildclient "github.com/openshift/origin/pkg/build/client"
-	buildcontrollerfactory "github.com/openshift/origin/pkg/build/controller/factory"
+	buildcontroller "github.com/openshift/origin/pkg/build/controller/build"
```
is there a compelling reason not to fold this whole file into build_controller.go?

csrwng replied:
this is part of @deads2k's recent change to the build controller. I believe this is a pattern we want to following going forward?

bparees replied:
hm.  it looks weird to me, but i'll ignore it for now.

deads2k replied:
Eventually we want to push this wiring down into sub-repos, but it wasn't a primary objective and we don't have a solid interface yet.  Right now, the it would create really strange interdependencies.

bparees commented on pkg/cmd/server/start/start_master.go:
```diff
@@ -754,7 +754,6 @@ func startControllers(oc *origin.MasterConfig, kc *kubernetes.MasterConfig) erro
 
 	// no special order
 	if configapi.IsBuildEnabled(&oc.Options) {
-		oc.RunBuildPodController()
```
i'd like to understand why this block is special, instead of handling the buildconfigchangecontroller in the for loop above.

csrwng replied:
the loop above is for the new style of controllers (where they live in an array and initialized via a function in that array) (ala kube). We have not yet changed the config change controller to initialize in that way.

bparees replied:
got it.

csrwng commented on pkg/build/controller/build/build_controller.go:
```diff
@@ -0,0 +1,821 @@
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: v1core.New(params.KubeClientExternal.Core().RESTClient()).Events("")})
+
+	buildListerUpdater := buildclient.NewOSClientBuildClient(params.OpenshiftClient)
+	buildDeleter := buildclient.NewBuildDeleter(params.OpenshiftClient)
```
Nope, we can certainly change that

csrwng commented:
@bparees please take a look at the last commit. I addressed your comments and changed the way we handle details updates from the pod to use annotations. I really don't like the potential for races on the build phase/reason/message. I'd rather have the build controller be the only one that can update those.

bparees commented on pkg/build/controller/build/build_controller.go:
```diff
@@ -296,11 +309,12 @@ func (bc *BuildController) cancelBuild(build *buildapi.Build) (*buildUpdate, err
 }
 
+// handleNewBuild will check whether policy allows running the new build and if so, creates a pod
+// for the build and returns and update to move it to the Pending phase
```
s/and/an/

bparees commented on pkg/build/controller/build/build_controller.go:
```diff
@@ -529,6 +547,16 @@ func (bc *BuildController) handleActiveBuild(build *buildapi.Build, willBeDroppe
+	}
+	if _, ok := build.Annotations[buildapi.BuildPodRequestedMessageAnnotation]; ok {
+		update.setMessage(build.Annotations[buildapi.BuildPodRequestedMessageAnnotation])
+	}
```
cool, i like this!

bparees commented on pkg/build/registry/build/strategy.go:
```diff
@@ -114,11 +114,13 @@ func (detailsStrategy) PrepareForUpdate(ctx apirequest.Context, obj, old runtime
+
+	newBuild.Annotations[api.BuildPodRequestedPhaseAnnotation] = string(phase)
+	newBuild.Annotations[api.BuildPodRequestedReasonAnnotation] = string(reason)
+	newBuild.Annotations[api.BuildPodRequestedMessageAnnotation] = message
```
godoc needs updating w/ this behavior change.

csrwng commented:
Updates made, new [annotation constants](https://github.com/openshift/origin/pull/14289/files#diff-f23c2970936ec5cfcb60d485ae96bcffR34) will require @openshift/api-review

csrwng commented:
[test]

csrwng commented:
[test]

csrwng commented:
@bparees still don't have a clean test. Latest run has one extended test failing... 
`Optimized image builds should succeed as an admin [Conformance]` 
there's also an integration build that failed, but succeeded in previous runs ... more like a flake 
`TestProjectLifecycle`

csrwng commented:
[testextended][extended:core(builds)]

csrwng commented:
Flakes https://github.com/openshift/origin/issues/14368 https://github.com/openshift/origin/issues/13980

csrwng commented:
Just realized I didn't tag the entire team... apologies. @openshift/devex ptal

csrwng commented:
@openshift/api-review bump

deads2k commented:
> Updates made, new annotation constants will require @openshift/api-review

Apologies if this was discussed elsewhere, but why are some of these fields moving to annotations?  If it were all the fields, I'd think you trying to separate spec from status (though given a separate subresource, I'm not sure that's necessary).  Were you having trouble stomping status accidentally somewhere?  That would seem to be an argument for computing  patch.

csrwng commented:
The problem is that we have racing updates. The builder pod can, in case of failures, set a message/reason/phase to a specific type of failure. This type of update is also in the realm of the build controller. Other updates such as revision built, time spent in each stage, etc. don't conflict with controller updates.

deads2k commented:
> The problem is that we have racing updates. The builder pod can, in case of failures, set a message/reason/phase to a specific type of failure. This type of update is also in the realm of the build controller. Other updates such as revision built, time spent in each stage, etc. don't conflict with controller updates.

Ok, so racing updates would normally be resolved by building patches:
 1. threadA gets build/z@rv6
 1. threadB gets build/z@rv6
 1. threadA makes patch for status/message
 1. threadB makes patch for status/message
 1. threadA wins
 1. threadB fails on conflict, re-gets with build/z@rv7 and can decide if he should overwrite the fields again.

bparees commented:
> Ok, so racing updates would normally be resolved by building patches:

it's not that kind of race.

it's two threads both trying to set a message on the object that indicates that thread has put the object into that state, however there are cases where if someone else already put the message there, you do NOT want to replace their message.  So the alternative implementation is to look at the current message and decide if it's one that should be overwritten or not.  This mechanism of handling prioritization feels cleaner.

deads2k commented:
> however there are cases where if someone else already put the message there, you do NOT want to replace their message. So the alternative implementation is to look at the current message and decide if it's one that should be overwritten or not.

Don't you already have to have that logic with this implementation too so that you can handle cases where the is set a split second earlier and you have decide whether to set the annotation?  And if you don't have it when you're setting the annotation, you'd have to have in a spot that observes the annotation was set and a message was set and the code has to decide if it should overwrite the message.

csrwng commented:
Rebased and updated the build strategy to prevent updates to the status and reason fields if the phase is already set to failed. Removed new annotation constants. Technically no longer needs API review.

csrwng commented:
[test]

smarterclayton commented:
Granted!  The null review is the best kind.  You have mastered the zen of
API changes and are ready to seek out your mountain.

On Jun 5, 2017, at 3:00 PM, Cesar Wong <notifications@github.com> wrote:

Rebased and updated the build strategy to prevent updates to the status and
reason fields if the phase is already set to failed. Removed new annotation
constants. Technically no longer needs API review.

—
You are receiving this because you are on a team that was mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/14289#issuecomment-306274125>,
or mute
the thread
<https://github.com/notifications/unsubscribe-auth/ABG_p_18Yngct_Y1bSTr8ZpCD6QUG03Eks5sBFA4gaJpZM4NjFNm>
.

csrwng commented:
Flake https://github.com/openshift/origin/issues/13829
[test]

csrwng commented:
[testextended][extended:core(builds)]

csrwng commented:
Changes requested by @jim-minter 
- [ ] Add description of build state machine at the beginning of the build_controller
- [ ] Explain the purpose of the buildUpdate object better at the beginning of that file
- [ ] Investigate removing the podDeletedKey by simply falling back to the REST client immediately when a pod is not found.
- [ ] Investigate using the generated informers and clients instead of the legacy ones

@jim-minter please let me know if there's something I missed.

jim-minter commented:
Looks good @csrwng; the other things I noted were:
- the state machine description should specifically mention that the builder pod currently updates phase, reason, message directly via the details subresource and that this is to be fixed
- ensure there's an issue or trello card to get the builder pod to update a condition instead of writing phase/reason/message directly
- we talked about upon build completion adding the next build directly to the controller queue rather than editing the BuildAcceptedAnnotation.  Suggest for now at least ensuring that there's an issue or trello card to look at this.

jim-minter commented on pkg/cmd/server/bootstrappolicy/controller_policy.go:
```diff
@@ -49,25 +49,13 @@ func init() {
-
-	// build-pod-controller
-	addControllerRole(rbac.ClusterRole{
-		ObjectMeta: metav1.ObjectMeta{Name: saRolePrefix + InfraBuildPodControllerServiceAccountName},
```
Can also delete InfraBuildPodControllerServiceAccountName const in infra_sa_policy.go

jim-minter commented on pkg/build/api/validation/validation.go:
```diff
@@ -752,3 +732,56 @@ func ValidateNodeSelector(nodeSelector map[string]string, fldPath *field.Path) f
+	return patch, nil
+}
+
+func ApplyBuildPatch(build *buildapi.Build, patch []byte) (*buildapi.Build, error) {
```
this is only called by build_controller_test.go - worth making it just a test function?

jim-minter commented on pkg/build/api/validation/validation.go:
```diff
@@ -752,3 +732,56 @@ func ValidateNodeSelector(nodeSelector map[string]string, fldPath *field.Path) f
+	if err != nil {
+		return nil, fmt.Errorf("error encoding older: %v", err)
+	}
+	patch, err := strategicpatch.CreateTwoWayMergePatch(olderJSON, newerJSON, &v1.Build{})
```
random passing comment, not asking for a change - I wonder if by using json patch (rfc6902) it would be possible to remove all this codec work - just work through the fields of buildUpdate and send it straight up [{"op": "replace", "path": "/path/to/field", "value": *field}, ...]

jim-minter replied:
actually, I wonder if doing that almost then get rids of the need for the buildUpdate object - what is updatable by the controller is just defined in the function which builds the json patch

csrwng replied:
It's still nice to have an in-memory representation of the update to determine what has been changed (for example, to know whether there is a phase transition)

jim-minter commented on pkg/build/controller/build/podcreationstrategy.go:
```diff
@@ -0,0 +1,46 @@
+package build
```
AFAICS this code has been copied from pkg/build/controller/factory/factory.go and the original needs to be deleted?

jim-minter commented on pkg/build/controller/factory/factory.go:
```diff
@@ -68,110 +65,6 @@ func limitedLogAndRetry(buildupdater buildclient.BuildUpdater, maxTimeout time.D
-		RetryManager: controller.NewQueueRetryManager(
-			queue,
-			cache.MetaNamespaceKeyFunc,
-			limitedLogAndRetry(factory.BuildUpdater, 30*time.Minute),
```
possibly can be more aggressive with code deletion here? for example, I think maybe limitedLogAndRetry, TestLimitedLogAndRetryFinish and TestLimitedLogAndRetryProcessing can be dispatched.  Check if there are more.  I'll be sorry to see `errors.New("funky error")` go though :)

jim-minter replied:
I might have gone deletion-happy but I think there's a mountain that can be trimmed here - candidates storeSyncedPollPeriod, keyListerGetter, podEnumerator, typeBasedFactoryStrategy (as already mentioned), podLW, buildDeleteLW, listPods (after the previous two are removed), newBuildLW, newImageStreamLW, ControllerClient.

I think maybe some of this is off the back of Clayton's prior commit too.

jim-minter commented on pkg/build/controller/strategy/custom.go:
```diff
@@ -113,6 +113,7 @@ func (bs *CustomBuildStrategy) CreateBuildPod(build *buildapi.Build) (*kapi.Pod,
 		setupDockerSocket(pod)
 		setupDockerSecrets(pod, build.Spec.Output.PushSecret, strategy.PullSecret, build.Spec.Source.Images)
 	}
+	setOwnerReference(pod, build)
```
there is quite a lot of overlap between the implemention of these three strategies - maybe a follow-up issue?

jim-minter commented on pkg/cmd/server/bootstrappolicy/controller_policy.go:
```diff
@@ -49,25 +49,13 @@ func init() {
 			rbac.NewRule("create").Groups(buildGroup, legacyBuildGroup).Resources("builds/optimizeddocker", "builds/docker", "builds/source", "builds/custom", "builds/jenkinspipeline").RuleOrDie(),
-			rbac.NewRule("get").Groups(imageGroup, legacyImageGroup).Resources("imagestreams").RuleOrDie(),
+			rbac.NewRule("get", "list").Groups(imageGroup, legacyImageGroup).Resources("imagestreams").RuleOrDie(),
+			rbac.NewRule("get", "list").Groups(kapiGroup).Resources("secrets").RuleOrDie(),
```
I'm curious - was this added because it didn't work without it?

csrwng replied:
no, it actually does work without it but we do access secrets (through the cache), so technically we should have permission to view them.

jim-minter replied:
@deads2k should policy rules be added when controllers access objects via caches even though the code works without them?  I was wondering what the preferred practice is when I was doing the template instance controller as well.

jim-minter commented on test/common/build/controllers.go:
```diff
@@ -653,6 +649,10 @@ func RunBuildRunningPodDeleteTest(t testingT, clusterAdminClient *client.Client,
 	}
+
+	// allow time for the failed event to get created
+	time.Sleep(5 * time.Second)
```
Is that a flake waiting to happen?

Also, above in this file:
```
	// give the poddeletecontroller's build cache time to be updated with the build object
	// so it doesn't get a miss when looking up the build while processing the pod delete event.
	time.Sleep(10 * time.Second)
```

Is the above one straight removable now the poddeletecontroller has gone?

csrwng replied:
So the reason the delay is added here is because we immediately check whether an event was produced when we detect that a build was updated through the watch. Previously, in some code paths (like this one), we would emit the event before doing the update. Now we consistently emit the event after the update happened. Adding the wait here was my quick way of allowing the event to be produced after the build update happened. If you have an idea of how to better handle it, I'm open to suggestions :)

csrwng replied:
> Is the above one straight removable now the poddeletecontroller has gone?

We still handle the pod delete event and update the build accordingly. The comment should be updated though.

jim-minter replied:
so I think the 10s sleep should be able to go now because findMissingPod is called.  The 5s sleep could at least be replaced by a wait.Poll.

bparees replied:
bump on this thread.  agree w/ @jim-minter that it looks like the 10s sleep can go.

for the 5s sleep, yeah a wait.Poll would be better, and i think we'd feel comfortable making it 30s or something to avoid flakes, too (Since we normally wouldn't wait the full 30s) then.

csrwng commented:
> ensure there's an issue or trello card to get the builder pod to update a condition instead of writing phase/reason/message directly

https://trello.com/c/NVTCdbKe/1045-13-conditions-for-builds-techdebtbuilds

csrwng commented:
> we talked about upon build completion adding the next build directly to the controller queue rather than editing the BuildAcceptedAnnotation. Suggest for now at least ensuring that there's an issue or trello card to look at this.

https://trello.com/c/mJ8qxqWY/1238-build-controller-investigate-having-the-policy-add-a-build-directly-to-the-queue-when-a-build-completes

csrwng commented:
> there is quite a lot of overlap between the implemention of these three strategies - maybe a follow-up issue?

https://github.com/openshift/origin/issues/14537

csrwng commented:
Created a Trello card to switch to the generated client. It looks like a little bit more work than I'd like to take on with this PR.
https://trello.com/c/X02R7Oi6/1239-use-generated-client-and-informers-in-build-buildconfig-imagetrigger-controllers

jim-minter commented on pkg/build/controller/build/build_controller.go:
```diff
@@ -0,0 +1,850 @@
+}
+
+// getPodByKey looks up a pod by key in the podInformer cache
+func (bc *BuildController) getPodByKey(key string) (*kapi.Pod, error) {
```
I think this function is unused and can be removed

csrwng replied:
will do

jim-minter commented on pkg/build/controller/build/build_controller.go:
```diff
@@ -0,0 +1,850 @@
+	obj, exists, err := bc.buildInformer.GetIndexer().GetByKey(key)
+	if err != nil {
+		glog.V(2).Infof("Unable to retrieve build %q from store: %v", key, err)
+		bc.queue.AddRateLimited(key)
```
maybe remove this, and call bc.handleError() in work() if bc.getBuildByKey() fails?

csrwng replied:
will fix

jim-minter commented on pkg/build/controller/build/build_controller.go:
```diff
@@ -0,0 +1,850 @@
+
+// podDeletedUpdate returns a buildUpdate object to apply to a build that has had
+// its pod deleted
+func podDeletedUpdate() *buildUpdate {
```
just a thought - is it worth calling this transitionToPhaseError() and accept the reason and message as parameters, then use this as a helper in the places where we transition to error?  It would be clearer than calling it podDeletedUpdate.

jim-minter replied:
or even transitionToPhase(phase, reason, message)

jim-minter replied:
I think doing this would further help audit/understand state transitions

jim-minter commented on pkg/build/controller/build/build_controller.go:
```diff
@@ -0,0 +1,850 @@
+		glog.V(4).Infof("Updating build %s -> %s%s", buildDesc(build), *update.phase, reasonText)
+	}
+
+	// Treat a failed->failed update as a state transition when the completionTime is getting
```
Comment under what circumstances this actually happens

csrwng replied:
will fix

jim-minter commented on pkg/build/controller/build/build_controller.go:
```diff
@@ -0,0 +1,850 @@
+
+	// Treat a failed->failed update as a state transition when the completionTime is getting
+	// updated. This will cause an event to be emitted and completion processing to trigger.
+	if !stateTransition && build.Status.Phase == buildapi.BuildPhaseFailed && update.completionTime != nil {
```
For intelligibility suggest pushing this up the function:
```golang
if update.phase != nil && (*update.phase) != build.Status.Phase {
	stateTransition = true
} else if build.Status.Phase == buildapi.BuildPhaseFailed && update.completionTime != nil {
	// comments
}

if stateTransition {
	// ...
}
```

csrwng replied:
will update

jim-minter commented on pkg/build/controller/build/build_controller.go:
```diff
@@ -0,0 +1,850 @@
+	}
+
+	// Ensure that a pod name annotation has been set on the build if a pod is available
+	if update.podNameAnnotation == nil && !common.HasBuildPodNameAnnotation(build) && pod != nil {
```
I think a) this block is superfluous, and b) if it isn't, I'm not convinced it belongs in this function

csrwng replied:
So we will set the podName annotation when we move the build from new to pending. However, if the update to pending failed, then we may not have pod name annotation on the build. This ensures that if there's a pod, we will have a pod name annotation. I can see that it can be moved to handleActiveBuild. Would that work?

jim-minter replied:
from irc: the problem is that createBuildPod is edge driven when it should be level driven; it should be retriable, assuming that a pre-existing pod which looks like a build pod is a build pod which it created, and that should be the sole place that we move from new->pending and do all the updates associated with that.  Then you can get rid of the hack at the start of handleNewBuild, and the "Ensure that a pod name annotation has been set on the build if a pod is available" in updateBuild()

csrwng replied:
@jim-minter I will open a card to investigate switching createBuildPod to be level driven. A long time ago we wanted to make all build handling level driven and at some point gave up. I don't remember what blocked us at the time, but it may have just been other priorities. @bparees should remember. I will punt on doing that here though given that I want to get this pull in sooner than later. It seems that maybe having a narrower focus such as the transition from new to pending would be easier.

bparees replied:
there were a lot of challenges in making builds level driven, one of them is that you no longer are guaranteed to get a build for each commit/image change, which for some people would be unacceptable/undesirable (though for others it's desirable, but that's why we have the SerialLatestOnly queue policy).

ultimately beyond the challenges to doing it, we dropped it because we got the behavior we wanted out of edge triggered builds.  Namely, introducing serial builds solved most of the problems we were trying to solve by moving to level driven builds.

jim-minter commented on pkg/build/controller/build/build_controller.go:
```diff
@@ -0,0 +1,850 @@
+		case buildapi.BuildPhaseComplete:
+			bc.recorder.Eventf(patchedBuild, kapi.EventTypeNormal, buildapi.BuildCompletedEventReason, fmt.Sprintf(buildapi.BuildCompletedEventMessage, patchedBuild.Namespace, patchedBuild.Name))
+		case buildapi.BuildPhaseError,
+			buildapi.BuildPhaseFailed:
```
insidious newline!

jim-minter commented on pkg/build/controller/build/build_controller.go:
```diff
@@ -0,0 +1,850 @@
+}
+
+// isTerminal returns true if the given build phase is a terminal state
+func isTerminal(phase buildapi.BuildPhase) bool {
```
this is identical to util.IsBuildComplete() - please combine
(I prefer the word terminal in the name, fwiw)

jim-minter replied:
I don't like that in some places "terminal" means (complete|failed|error|cancelled), and in some places it means the same minus (complete/failed && completiontimestamp == nil).  Seems like a recipe for logic errors.  Is it possible to audit and improve this at all now?

csrwng replied:
will move to buildutil and use one function ... we take the completiontimestamp into consideration only when deciding to ignore the build or not. Not sure that changing things there would make it clearer.

jim-minter commented on pkg/build/controller/build/build_controller.go:
```diff
@@ -0,0 +1,850 @@
+	// means that the build is active and its status should be updated
+	if podErr == nil && pod != nil {
+		if !pod.CreationTimestamp.Before(build.CreationTimestamp) {
+			return bc.handleActiveBuild(build, pod, podErr)
```
I half think that for long term maintainability it might be nicer to update phase and requeue at ths point

csrwng replied:
hmm...  that's what we're going to do (update the phase) by sending it to handleActiveBuild

jim-minter commented on pkg/build/controller/build/build_controller.go:
```diff
@@ -0,0 +1,850 @@
+	// If a pod was found, and it was created after the build was created, it
+	// means that the build is active and its status should be updated
+	if podErr == nil && pod != nil {
+		if !pod.CreationTimestamp.Before(build.CreationTimestamp) {
```
I wonder if the error handling in createBuildPod can be brought up to here so that the existing pod case is handled in one place, whether older than the build object or not.

jim-minter commented on pkg/build/controller/build/build_controller.go:
```diff
@@ -0,0 +1,850 @@
+// updateBuild is the single place where any update to a build is done in the build controller.
+// It will check that the update is valid, peform any necessary processing such as calling HandleBuildCompletion,
+// and apply the buildUpdate object as a patch.
+func (bc *BuildController) updateBuild(build *buildapi.Build, update *buildUpdate, pod *kapi.Pod) {
```
why does updateBuild not return error?  If the patch fails, shouldn't a requeue happen?

csrwng replied:
It should return an error, will fix

jim-minter commented on pkg/build/controller/build/build_controller.go:
```diff
@@ -0,0 +1,850 @@
+	}
+
+	// If there was some other error retrieving the pod, there may be an issue with the API server, better retry
+	if !errors.IsNotFound(podErr) {
```
push up to handleBuild()?

csrwng replied:
actually will remove this ... the only error that can be returned by the cache store is NotFound

csrwng replied:
hmm, will leave a check in check that changes in the future, but yes, will handle in the calling function

jim-minter commented on pkg/build/controller/build/build_controller.go:
```diff
@@ -0,0 +1,850 @@
+
+	runPolicy := policy.ForBuild(build, bc.runPolicies)
+	if runPolicy == nil {
+		return nil, fmt.Errorf("unable to determine build policy for %s", buildDesc(build))
```
does this justify moving to the Error state?

csrwng replied:
If we run into this error, it likely means that there's something on the build object that we don't understand (for example, in a new version of the master we've introduced a new policy type). I don't think we should do anything to the build.

jim-minter commented on pkg/build/controller/build/build_controller.go:
```diff
@@ -0,0 +1,850 @@
+		return
+	}
+
+	if strategy.IsFatal(err) {
```
I think this isn't right.  FWIW I think we may be underusing the FatalError concept, but more importantly, I think if one bubbles up to here, then it should result in a transition to Error state.

csrwng replied:
This one needs a little more thought. Will leave alone for now and create an issue to follow up.

jim-minter commented on pkg/build/controller/build/build_controller.go:
```diff
@@ -0,0 +1,850 @@
+	// Set the output Docker image reference.
+	ref, err := bc.resolveOutputDockerImageReference(build)
+	if err != nil {
+		update.setReason(buildapi.StatusReasonInvalidOutputReference)
```
No phase change here?

jim-minter replied:
Quite possibly legitimate, but warrants a comment

csrwng replied:
Yes, if the output image stream doesn't exist, then we should not start the build

jim-minter commented on pkg/build/controller/build/build_controller.go:
```diff
@@ -0,0 +1,850 @@
+	// Create the build pod spec
+	buildPod, err := bc.createPodSpec(build, ref)
+	if err != nil {
+		update.setReason(buildapi.StatusReasonCannotCreateBuildPodSpec)
```
No phase change here?

csrwng replied:
An error here would mean that there's something in the build object that we don't understand such as a strategy we don't recognize. I would think that the safer action would be to not change the phase here. @bparees thoughts?

bparees replied:
what phase would we change it to?  our model is that the "pending" phase means we've successfully created the pod.

csrwng replied:
the alternative would be to change it to the error phase

bparees replied:
if the question is "should we just immediately mark the build failed at this point" the answer is no, not unless we're going to start sorting through what the error was.  The pod create call could have failed for any number of temporary reasons, so we need to retry it.

csrwng replied:
The createPodSpec is not telling the rest api to create a pod on the master. It's simply constructing a pod object in memory that corresponds to the build that we passed it. Like I said, the main reason it would fail is if the build has a strategy that we don't understand.

jim-minter commented on pkg/build/controller/build/build_controller.go:
```diff
@@ -0,0 +1,850 @@
+// enqueueBuildForPod adds the build corresponding to the given pod to the controller
+// queue. If a build is not found for the pod, then an error is logged.
+func (bc *BuildController) enqueueBuildForPod(pod *kapi.Pod) {
+	buildName := buildutil.GetBuildName(pod)
```
can't this entire function just be `bc.queue.Add(pod.Namespace + "/" + buildutil.GetBuildName(pod))` ?

csrwng replied:
yup, will update

csrwng commented:
@bparees @jim-minter I am done with changes and unit tests. The code should be ready for a final review.

bparees commented on pkg/build/controller/build/doc.go:
```diff
@@ -0,0 +1,70 @@
+// The BuildController is responsible for implementing a build's lifecycle,
```
this is fantastic......now to remember it's here.

bparees commented on pkg/build/controller/build/doc.go:
```diff
@@ -0,0 +1,70 @@
+// push secret exists, and if not, it will update the build reason and message
+// with that information.
+//
+// Running - a build is update to this state if the corresponding pod is in the
```
s/update/updated/ (sorry)

csrwng commented on pkg/build/controller/build/build_controller.go:
```diff
@@ -0,0 +1,824 @@
+		// A scenario where this would happen is that we've introduced a
+		// new build strategy in the master, but the old version of the controller
+		// is still running. We don't want the old controller to move the
+		// build to the error phase and we don't want it to keep actively retrying.
```
Yes it will cause another loop through here when we update it, and we will see the build again on a relist. So not ideal.

csrwng commented on pkg/build/controller/build/build_controller.go:
```diff
@@ -0,0 +1,824 @@
+
+	patchedBuild, err := bc.patchBuild(build, update)
+	if err != nil {
+		if stateTransition && *update.phase == buildapi.BuildPhasePending {
```
because if we transitioned to pending, then we just created a pod, and at least in the original code, we don't retry the build in New state. The danger is that the new pod doesn't make it into the cache before the retry and we end up creating a new pod when retrying the build. Once the pod is created, we can catch up to the latest state by just observing the state of the pod. This is the transition that @jim-minter had suggested to be level-driven instead of edge-driven, which warrants further investigation.

bparees replied:
ok so we explicitly don't want to retry this particular event due to this particular failure because it could lead to us creating 2 pods for the same build.  got it.  I think it definitely merits a code comment, pending whether we rework things or not.

bparees replied:
@csrwng still looking for a comment in the code on this logic.  it's a subtle condition that i don't think will be immediately obvious to the next person touching this code.

csrwng replied:
@bparees thinking about this some more, I think a retry would be harmless, given that at worst we'll get a conflict trying to recreate the pod. Otherwise, the build should get updated properly based on the state of the pod. If you're ok with it, I'll remove the check and always return the error.

bparees replied:
@csrwng as long as we're handling the "pod already exists" error on create in a sane way, that sounds fine to me, and i certainly like the idea of removing the special case logic.

csrwng commented:
thx @bparees latest comments addressed

bparees commented:
add the code comment and i think it's ready for merge.  feel free to put the tag on yourself after you add the comment.

csrwng commented:
Removed special case for pending transition 
[merge]

mfojtik commented on pkg/build/controller/strategy/util.go:
```diff
@@ -216,3 +218,16 @@ func getContainerVerbosity(containerEnv []kapi.EnvVar) (verbosity string) {
+	t := true
+	pod.OwnerReferences = []metav1.OwnerReference{
+		{
+			APIVersion: "v1",
```
must be build.openshift.io/v1
`

bparees replied:
@mfojtik noted that this needs to be buildapiv1.SchemeGroupVersion

mfojtik replied:
also consider something like:

https://github.com/openshift/origin/pull/14582/files#diff-d09310d3307d040cedbddcd5fa308067R30

csrwng commented:
Thx @mfojtik and @bparees. Code updated.

csrwng commented:
rebased and fixed compile error with ImageStream informer

csrwng commented:
Fixed unit test to use generated informer for ImageStreams

csrwng commented:
Flake https://github.com/openshift/origin/issues/13984 on extended test

csrwng commented:
[testextended][extended:core(builds)]

smarterclayton commented:
[severity:bug]

smarterclayton commented:
[test]

smarterclayton commented:
[merge]
